resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ResourceTypeHelper.scala

```diff
@@ -0,0 +1,180 @@
+
+      try {
+        val resInfoClass = Utils.classForName(
+          "org.apache.hadoop.yarn.api.records.ResourceInformation")
+        val setResourceInformationMethod =
+          resource.getClass.getMethod("setResourceInformation", classOf[String],
+            resInfoClass)
+
+        val resourceInformation =
+          createResourceInformation(resourceName, amount, unit, resInfoClass)
+        setResourceInformationMethod.invoke(resource, resourceName, resourceInformation)
+      } catch {
+        case e: InvocationTargetException =>
+          if (e.getCause != null) {
+            throw e.getCause
+          } else {
+            throw e
+          }
+        case NonFatal(e) =>
+          logWarning(resourceTypesNotAvailableErrorMessage, e)
+      }
+    }
+    resource
+  }
+
+  def getCustomResourcesAsStrings(resource: Resource): String = {
+    if (resource == null) {
+      throw new IllegalArgumentException("Resource parameter should not be null!")
+    }
+
+    if (!ResourceTypeHelper.isYarnResourceTypesAvailable()) {
+      logWarning(resourceTypesNotAvailableErrorMessage)
+      return ""
+    }
+
+    var res: String = ""
+    try {
+      val resUtilsClass = Utils.classForName(
+        "org.apache.hadoop.yarn.util.resource.ResourceUtils")
+      val getNumberOfResourceTypesMethod = resUtilsClass.getMethod("getNumberOfKnownResourceTypes")
+      val numberOfResourceTypes: Int = getNumberOfResourceTypesMethod.invoke(null).asInstanceOf[Int]
+      val resourceClass = Utils.classForName(
+        "org.apache.hadoop.yarn.api.records.Resource")
+
+      // skip memory and vcores (index 0 and 1)
+      for (i <- 2 until numberOfResourceTypes) {
+        val getResourceInfoMethod = resourceClass.getMethod("getResourceInformation",
+          classOf[Int])
+        res ++= getResourceInfoMethod.invoke(resource, i.asInstanceOf[AnyRef]).toString()
+      }
+    } catch {
+      case e: InvocationTargetException =>
+        if (e.getCause != null) {
+          throw e.getCause
+        } else {
+          throw e
+        }
+      case NonFatal(e) =>
+        logWarning(resourceTypesNotAvailableErrorMessage, e)
+
+    }
+    res
+  }
+
+  def getAmountAndUnit(s: String): (Long, String) = {
+    try {
+      val AMOUNT_AND_UNIT_REGEX(amount, unit) = s
+      (amount.toLong, convertToYarnResourceFormat(unit))
+    } catch {
+      case _: MatchError => throw new IllegalArgumentException(
+        s"Value of resource type should match pattern $AMOUNT_AND_UNIT_REGEX, unmatched value: $s")
+    }
+  }
+
+  def convertToYarnResourceFormat(unit: String): String = {
+    unit match {
+      case "m" => "M"
+      case "g" => "G"
+      case "t" => "T"
+      case "p" => "P"
+      case _ => unit
+    }
+  }
+
+  private def createResourceInformation(
+      resourceName: String,
+      amount: Long,
+      unit: String,
+      resInfoClass: Class[_]) = {
```

Comment:
Missing return type.

Reply:
How I would specify return type here? As I cannot refer to the ResourceInformation class directly since by default Spark is compiled against Hadoop2.x and not Hadoop 3.
In Hadoop2.x we don't have that class in the YARN API jars.

Reply:
Use `Any` or something along those lines.

Reply:
Fixed